File: sections/62_codeforces_course_ladder/problems/31_rotate_columns/solution.py
import sys


def build_spf(limit: int) -> list[int]:
    spf = list(range(limit + 1))
    if limit >= 1:
        spf[1] = 1
    for prime in range(2, int(limit**0.5) + 1):
        if spf[prime] == prime:
            for multiple in range(prime * prime, limit + 1, prime):
                if spf[multiple] == multiple:
                    spf[multiple] = prime
    return spf


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    values = data[1:1 + data[0]]
    spf = build_spf(max(values))
    out = []
    for value in values:
        remaining = value
        multiplier = 1
        while remaining > 1:
            prime = spf[remaining]
            odd_exponent = False
            while remaining % prime == 0:
                odd_exponent = not odd_exponent
                remaining //= prime
            if odd_exponent:
                multiplier *= prime
        out.append(str(multiplier))
    print("\n".join(out))


if __name__ == "__main__":
    main()

File: sections/62_codeforces_course_ladder/problems/31_rotate_columns/solution.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> build_spf(int limit) {
    vector<int> spf(limit + 1);
    iota(spf.begin(), spf.end(), 0);
    if (limit >= 1) spf[1] = 1;
    for (long long p = 2; p * p <= limit; ++p) {
        if (spf[p] == p) {
            for (long long multiple = p * p; multiple <= limit; multiple += p) {
                if (spf[multiple] == multiple) spf[multiple] = p;
            }
        }
    }
    return spf;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int q;
    if (!(cin >> q)) return 0;
    vector<int> values(q);
    for (int &value : values) cin >> value;
    int limit = *max_element(values.begin(), values.end());
    vector<int> spf = build_spf(limit);

    for (int value : values) {
        int remaining = value;
        long long multiplier = 1;
        while (remaining > 1) {
            int prime = spf[remaining];
            bool odd_exponent = false;
            while (remaining % prime == 0) {
                odd_exponent = !odd_exponent;
                remaining /= prime;
            }
            if (odd_exponent) multiplier *= prime;
        }
        cout << multiplier << '\n';
    }
    return 0;
}
